core/src/main/java/com/graphhopper/routing/weighting/custom/CustomModelParser.java

```diff
@@ -405,40 +408,79 @@ private static SimpleCompiler createCompiler(long counter, Java.AbstractCompilat
+        return result;
+    }
+
+    static double getMaxSpeedForBlock(List<Statement> block, final double maxSpeed) {
+        if (block.isEmpty() || !IF.equals(block.get(0).getKeyword()))
+            throw new IllegalArgumentException("Every block must start with an if-statement");
+        if (block.get(0).getCondition().trim().equals("true"))
+            // this if statement is always executed while the other statements are never executed
+            // -> we just apply this one statement
+            return applyOperator(maxSpeed, block.get(0).getOperation(), block.get(0).getValue());
 
-                if (globalMin_maxSpeed <= 0)
-                    throw new IllegalArgumentException("speed is always limited to 0. This must not be but results from " + statement);
+        double blockMax = block.stream()
+                .mapToDouble(branch -> applyOperator(maxSpeed, branch.getOperation(), branch.getValue()))
+                .max()
+                .orElse(maxSpeed);
+        // if there is no 'else' statement it's like there is a 'neutral' branch that leaves the initial max speed as is
+        if (block.stream().noneMatch(st -> ELSE.equals(st.getKeyword())))
+            blockMax = Math.max(blockMax, maxSpeed);
+        return blockMax;
+    }
+
+    private static double applyOperator(double startValue, Statement.Op op, double applyValue) {
+        if (LIMIT.equals(op))
+            return Math.min(applyValue, startValue);
+        else if (Statement.Op.MULTIPLY.equals(op))
+            return startValue * applyValue;
+        else
+            throw new IllegalArgumentException("Unknown operator " + op.getName());
+    }
+
+    /**
+     * Splits a list of if, else_if, else, if, else_if, else_if, ... statements into if/else_if/else blocks
+     */
+    static List<List<Statement>> splitIntoBlocks(List<Statement> statements) {
+        if (!statements.isEmpty() && !IF.equals(statements.get(0).getKeyword()))
+            throw new IllegalArgumentException("The first statement must be an if-statement, but was: " + statements.get(0).getKeyword().getName());
+        List<List<Statement>> result = new ArrayList<>();
+        List<Statement> block = null;
+        Statement.Keyword prev = null;
+        for (Statement st : statements) {
+            final Statement.Keyword keyword = st.getKeyword();
+            if (ELSEIF.equals(keyword) && ELSE.equals(prev))
+                throw new IllegalArgumentException("After each else-statement there must be an if-statement");
+            if (ELSE.equals(keyword) && ELSE.equals(prev))
+                throw new IllegalArgumentException("After each else-statement there must be an if-statement");
+            if (IF.equals(keyword)) {
+                block = new ArrayList<>();
+                result.add(block);
             }
+            block.add(st);
+            prev = keyword;
         }
-
-        return globalMin_maxSpeed;
+        return result;
```

Comment:
Not sure if we should do this. We would duplicate the work of the compiler. There is no need for this code to fail here early as it would fail later.

Reply:
Ok, I thought it makes it more clear to first split the statements into blocks and then find the max per block instead of doing everything in one loop. It's true we do not necessarily need to check the order of else/else_if here, but I thought this way we get a more useful error message. The downside of course would be false positives in case there is still a bug here. Don't know, please feel free to adjust the code.

Reply:
Would you be ok with the latest commit?

Reply:
Sure, go ahead :)